docs/user-guide/setting-up-ci-cd.rst

```diff
@@ -0,0 +1,207 @@
+
+To ensure the test will fail if any test conversation fails, add 
+the ``--fail-on-prediction-errors`` flag:
+
+Note: End-to-end testing does **not** execute your action code. You will need to
+:ref:`test your action code <testing-action-code>` in a seperate step.
+
+NLU Comparison
+##############
+
+If you've made significant changes to your NLU training data (such as adding or
+splitting intents, or just adding/changing a lot of examples), you should run a
+:ref:`full NLU evaluation <nlu-evaluation>`. You'll want to compare
+the performance of the NLU model without your changes to an NLU model with your
+changes. 
+
+You can do this by running NLU testing in cross-validation mode:
+
+.. code-block:: bash
+
+   rasa test nlu --cross-validation
+
+or by training a model on a training set and testing it on a test set. If you use the train-test
+set approach, it is best to :ref:`shuffle and split your data <train-test-split>` using ``rasa data split`` every time, as
+opposed to using a static NLU test set, which can easily become outdated. 
+
+Since NLU comparison can be a fairly resource intensive test, you can set this
+test to run only when a certain tag (e.g. "NLU testing required") is present,
+or only when changes to NLU data or the NLU pipeline were made. 
+
+.. _testing-action-code:
+
+Testing Action Code
+###################
+
+The approach used to test your action code will depend on how it is
+implemented. Whichever method of testing your code you choose, you should
+include running those tests in your CI pipeline as well. For example, if you
+connect to external API's it is recommended to write unit tests to ensure 
+that those APIs respond as expected to common inputs.
+
+Continuous Deployment
+---------------------
+
+To get improvements out to your users frequently, you need to automate as
+much of the deployment process as possible. 
+
+CD steps usually run on push or merge to a certain branch, once CI checks have
+succeeded.
+
+.. contents::
+   :local:
+
+.. _uploading-a-model:
+
+Deploying your Rasa Model
+#########################
+
+You should already have a trained model from running end-to-end testing in your
+CI pipeline. You can set up your pipeline to upload the trained model to your
+Rasa server. If you're using Rasa X, you can also 
+make an `API call <https://rasa.com/docs/rasa-x/api/rasa-x-http-api/#tag/Models/paths/~1projects~1{project_id}~1models~1{model}~1tags~1{tag}/put>`_ 
+to tag the uploaded model as ``production`` (or whichever `deployment environment <https://rasa.com/docs/rasa-x/enterprise/deployment-environments/#>`_ you want
+to deploy it to).
+
+For Rasa X, this could look something like:
+
+.. code-block:: bash
+
+   curl -k -F "model=@models/my_model.tar.gz" "https://example.rasa.com/api/projects/default/models?api_token={your_api_token}"
+   curl -X PUT "https://example.rasa.com/api/projects/default/models/my_model/tags/production"
```

Comment:
Let's keep these as two separate ones bc they might not necessarily always go together (as you mention below), so

You can upload the model automatically, for example in Rasa X this might look like:

code

If using Rasa X you can also tag it automatically as production:

code